x-pack/platform/plugins/shared/fleet/server/services/agent_policies/package_policies_to_agent_inputs.ts

```diff
@@ -138,6 +142,12 @@ export const getFullInputStreams = (
                 data_stream: {
                   ...stream.data_stream,
                   ...compiledDataStream,
+                  ...(stream.vars?.[DATA_STREAM_TYPE_VAR_NAME]?.value
+                    ? { type: stream.vars?.[DATA_STREAM_TYPE_VAR_NAME]?.value }
+                    : {}),
+                  ...(stream.vars?.[DATASET_VAR_NAME]?.value
+                    ? { dataset: stream.vars?.[DATASET_VAR_NAME]?.value }
+                    : {}),
```

Comment:
At this point:
* `stream.data_stream` contains the default values.
* `compiledDataStream` contains the values in the compiled data stream, rendered from the agent template.

So if the compiled data stream doesn't include the `data_stream` object, custom type and dataset are lost.